core/src/main/scala/akka/kafka/internal/TransactionalSource.scala

```diff
@@ -31,49 +31,21 @@ private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerS
     extends KafkaSourceStage[K, V, TransactionalMessage[K, V]](
       s"TransactionalSource ${subscription.renderStageAttribute}"
     ) {
-  import TransactionalSource._
 
   require(consumerSettings.properties(ConsumerConfig.GROUP_ID_CONFIG).nonEmpty, "You must define a Consumer group.id.")
 
-  type Offset = Long
-
-  private trait InFlightRecords {
-    // Assumes that offsets per topic partition are added in the increasing order
-    // The assumption is true for Kafka consumer that guarantees that elements are emitted
-    // per partition in offset-increasing order.
-    def add(offsets: Map[TopicPartition, Offset]): Unit
-    def committed(offsets: Map[TopicPartition, Offset]): Unit
-    def revoke(revokedTps: Set[TopicPartition]): Unit
-    def reset(): Unit
-
-    def empty(): Boolean
+  override protected def logic(shape: SourceShape[TransactionalMessage[K, V]]): GraphStageLogic with Control = {
+    val txConsumerSettings = TransactionalSource.setIsolationLevel(consumerSettings)
+    new TransactionalSourceLogic(shape, txConsumerSettings, subscription) with TransactionalMessageBuilder[K, V] {
```

Comment:
To be able to mix in another message builder, I split the `TransactionalSourceLogic` similar to what `SingleSourceLogice` does.